cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,661 @@
+
+  key_map.insert_if(iter,
+                    iter + num_nodes,
+                    thrust::counting_iterator<size_type>(0),  // stencil
+                    is_field_name_node,
+                    d_hasher,
+                    d_equal,
+                    stream.value());
+  auto get_hash_value =
+    [key_map = key_map.get_device_view(), d_hasher, d_equal] __device__(auto node_id) -> size_type {
+    auto it = key_map.find(node_id, d_hasher, d_equal);
+    return (it == key_map.end()) ? size_type{0} : it->second.load();
+  };
+  // convert field nodes to node indices, and other nodes to enum value.
+  rmm::device_uvector<size_type> node_type(num_nodes, stream);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   node_type.begin(),
+                   node_type.end(),
+                   [node_categories = d_tree.node_categories.data(),
+                    is_field_name_node,
+                    get_hash_value] __device__(auto node_id) -> size_type {
+                     if (is_field_name_node(node_id))
+                       return static_cast<size_type>(NUM_NODE_CLASSES) + get_hash_value(node_id);
+                     else
+                       return static_cast<size_type>(node_categories[node_id]);
+                   });
+// TODO delete hash_map after node_type creation
+#ifdef NJP_DEBUG_PRINT
+  print_vec(cudf::detail::make_std_vector_async(node_type, stream), "node_type");
+#endif
+  // 2. Preprocessing: Translate parent node ids after sorting by level.
+  //   a. sort by level
+  //   b. get gather map of sorted indices
+  //   c. translate parent_node_ids to sorted indices
+  rmm::device_uvector<size_type> scatter_indices(num_nodes, stream);
+  thrust::sequence(rmm::exec_policy(stream), scatter_indices.begin(), scatter_indices.end());
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(cudf::detail::make_std_vector_async(scatter_indices, stream), "gpu.node_id");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.parent_node_ids, stream),
+            "gpu.parent_node_ids");
+  print_vec(cudf::detail::make_std_vector_async(node_type, stream), "gpu.node_type");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.node_levels, stream), "gpu.node_levels");
+#endif
+  rmm::device_uvector<NodeIndexT> parent_node_ids(d_tree.parent_node_ids, stream);  // make a copy
+  auto out_pid =
+    thrust::make_zip_iterator(scatter_indices.data(), parent_node_ids.data(), node_type.data());
+  // Uses cub radix sort.
+  thrust::stable_sort_by_key(rmm::exec_policy(stream),
+                             d_tree.node_levels.data(),
+                             d_tree.node_levels.data() + num_nodes,
+                             out_pid);
+  auto gather_indices = cudf::detail::scatter_to_gather(
+    scatter_indices.begin(), scatter_indices.end(), num_nodes, stream);
+
+  rmm::device_uvector<NodeIndexT> parent_indices(num_nodes, stream);
+  *thrust::device_pointer_cast(parent_indices.data()) = -1;
```

Comment:
Can we use `set_element_async` here to avoid the extra synchronization?
```suggestion
  parent_indices.set_element_async(0, no_parent_sentinel, stream)
```

Reply:
needed to replace this with `thrust::fill` for json-lines support.